07-MemoryManagementKernelModule/allocation.c

```diff
@@ -0,0 +1,323 @@
+			ptr = kmalloc(size, GFP_KERNEL);
+			getnstimeofday(&t2);
+			deltaAlloc[i] = sub(t2, t1);
+
+			getnstimeofday(&t1);
+			kfree(ptr);
+			getnstimeofday(&t2);
+			deltaFree[i] = sub(t2, t1);
+		}
+
+		long avrAllocNS = 0, avrFreeNS = 0;
+		long minAllocNS = 0, minFreeNS = 0;
+		long maxAllocNS = 0, maxFreeNS = 0;
+		long deltaAllocNS[attempts], deltaFreeNS[attempts];
+
+		i = 0;
+		for (; i < attempts; i++) {
+			deltaAllocNS[i] = deltaAlloc[i].tv_sec * 1000000000;
+			deltaAllocNS[i] += deltaAlloc[i].tv_nsec;
+			deltaFreeNS[i] = deltaFree[i].tv_sec * 1000000000;
+			deltaFreeNS[i] += deltaFree[i].tv_nsec;
+		}
+
+		minAllocNS = maxAllocNS = deltaAllocNS[0];
+		minFreeNS = maxFreeNS = deltaFreeNS[0];
+		i = 0;
+		for (; i < attempts; i++) {
+			avrAllocNS += deltaAllocNS[i];
+			avrFreeNS += deltaFreeNS[i];
+			minAllocNS = min(minAllocNS, deltaAllocNS[i]);
+			minFreeNS = min(minFreeNS, deltaFreeNS[i]);
+			maxAllocNS = max(maxAllocNS, deltaAllocNS[i]);
+			maxFreeNS = max(maxFreeNS, deltaFreeNS[i]);
+		}
+
+		avrAllocNS /= attempts;
+		avrFreeNS /= attempts;
+
+		pr_info("kmalloc:2^%02d+%d	%09ld	%09ld	%09ld	%09ld	%09ld	%09ld",
+			power, oversize, minAllocNS, avrAllocNS, maxAllocNS,
+			minFreeNS, avrFreeNS, maxFreeNS);
+
+		oversize++;
+		oversize %= 2;
+		if (oversize == 0)
+			power++;
+	}
+
+	power = 0;
+	oversize = 0;
+
+	pr_info("kzalloc:size	minAlloc	avrAlloc	maxAlloc	minFree		avrFree		maxFree");
+
+	while (power < 65) {
+		const int attempts = 9;
+		struct timespec t1, t2;
+		struct timespec deltaAlloc[attempts], deltaFree[attempts];
+		unsigned long long size;
+
+		int i = 0;
+
+		for (; i < attempts; i++) {
+			size = (1 << power) + oversize;
+			getnstimeofday(&t1);
+			ptr = kzalloc(size, GFP_KERNEL);
+			getnstimeofday(&t2);
+			deltaAlloc[i] = sub(t2, t1);
+
+			getnstimeofday(&t1);
+			kfree(ptr);
+			getnstimeofday(&t2);
+			deltaFree[i] = sub(t2, t1);
+		}
+
+		long avrAllocNS = 0, avrFreeNS = 0;
+		long minAllocNS = 0, minFreeNS = 0;
+		long maxAllocNS = 0, maxFreeNS = 0;
+		long deltaAllocNS[attempts], deltaFreeNS[attempts];
+
+		i = 0;
+		for (; i < attempts; i++) {
+			deltaAllocNS[i] = deltaAlloc[i].tv_sec * 1000000000;
+			deltaAllocNS[i] += deltaAlloc[i].tv_nsec;
+			deltaFreeNS[i] = deltaFree[i].tv_sec * 1000000000;
+			deltaFreeNS[i] += deltaFree[i].tv_nsec;
+		}
+
+		minAllocNS = maxAllocNS = deltaAllocNS[0];
+		minFreeNS = maxFreeNS = deltaFreeNS[0];
+		i = 0;
+		for (; i < attempts; i++) {
+			avrAllocNS += deltaAllocNS[i];
+			avrFreeNS += deltaFreeNS[i];
+			minAllocNS = min(minAllocNS, deltaAllocNS[i]);
+			minFreeNS = min(minFreeNS, deltaFreeNS[i]);
+			maxAllocNS = max(maxAllocNS, deltaAllocNS[i]);
+			maxFreeNS = max(maxFreeNS, deltaFreeNS[i]);
+		}
+
+		avrAllocNS /= attempts;
+		avrFreeNS /= attempts;
+
+		pr_info("kzalloc:2^%02d+%d	%09ld	%09ld	%09ld	%09ld	%09ld	%09ld",
+			power, oversize, minAllocNS, avrAllocNS, maxAllocNS,
+			minFreeNS, avrFreeNS, maxFreeNS);
+
+		oversize++;
+		oversize %= 2;
```

Comment:
In my opinion one line for oversize update is enough. Like oversize = 1 - oversize
or oversize ^= 1